jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -157,35 +179,21 @@ public Result find() {
       return Optional.empty();
     }
 
-    ClassPool classPool = new ClassPool();
-    classPool.appendSystemPath();
-
-    try {
-      CtClass[] mainMethodParams = new CtClass[] {classPool.get("java.lang.String[]")};
-
-      try (InputStream classFileInputStream = Files.newInputStream(file)) {
-        CtClass fileClass = classPool.makeClass(classFileInputStream);
-
-        // Check if class contains 'psvm' (see class javadoc).
-        CtMethod mainMethod = fileClass.getDeclaredMethod("main", mainMethodParams);
-
-        if (CtClass.voidType.equals(mainMethod.getReturnType())
-            && Modifier.isStatic(mainMethod.getModifiers())
-            && Modifier.isPublic(mainMethod.getModifiers())) {
-          return Optional.of(fileClass.getName());
-        }
-
-      } catch (NotFoundException ex) {
-        // Ignores main method not found.
-
-      } catch (IOException ex) {
-        // Could not read class file.
-        eventDispatcher.dispatch(LogEvent.warn("Could not read file: " + file));
+    MainClassVisitor mainClassVisitor = new MainClassVisitor();
+    try (InputStream classFileInputStream = Files.newInputStream(file)) {
+      ClassReader reader = new ClassReader(classFileInputStream);
+      reader.accept(mainClassVisitor, 0);
+      if (mainClassVisitor.visitedMainClass) {
+        return Optional.of(reader.getClassName().replace("/", "."));
       }
 
-    } catch (NotFoundException ex) {
-      // Thrown if 'java.lang.String' is not found in classPool.
-      throw new RuntimeException(ex);
+    } catch (ArrayIndexOutOfBoundsException ignored) {
```

Comment:
Having trouble finding any, but I'll keep looking and let you know.

Reply:
I'd be curious to understand how this would trigger? Does the mean the `.class`es we are inspecting are potentially corrupt?

Reply:
This exception was thrown during one of our tests (`testFindMainClass_simple`, while trying to load `NotAClassButMightBe.class`). `ClassReader` does a bunch of stuff by indexing into a byte array without checking, and `NotAClassButMightBe.class` is just an empty file, so it just ends up going out of bounds.

Reply:
Normally I would except these kind of exceptions to slip through directly to the user, but this would probably be a difficult error to debug on the user side. It's nice for us to detect these issues, but it might actually be okay to fail here. We don't want to be packaging bad class files anyway.

Reply:
I don't think we should fail the build, I think it should behave the same way it would if it found a non-class file, but maybe we should log a warning instead of a debug message?

I don't even know if it's up to us to decide if their project files are valid or not, at least not at this stage where it's just trying to find the main class.

Reply:
Sounds good. Might make sense to move the catch to exactly the method we expect to throw it, so we are isolating for that single case.

Reply:
Just a final thing: It's not obviously clear to me which method call is throwing this exception as a developer, as it's not exposed by the API. Isolating it to the exact method call with a separate try block would clarify that. It does make the logic of this block a little more complicated though.

Reply:
Yeah, it's `ClassReader`. And I think it's already isolated to the degree it should be. It seems like all 3 places the `ClassReader` is used do some unchecked array indexing. I can add a comment though.